kvevent: refactor blocking buffer to use chunked linked list

jayshrivastava opened the pull request:
This change updates `kvevent.blockingBuffer` to use a chunked linked list instead of a regular linked list. It also updates the benchmark we use to measure performance for this struct. Please see the commit messages for more info.

Results:
```
name          old time/op    new time/op    delta
MemBuffer-10     377ms ± 1%     511ms ± 0%  +35.54%  (p=0.000 n=8+9)

name          old alloc/op   new alloc/op   delta
MemBuffer-10    33.6MB ± 0%     0.4MB ± 3%  -98.93%  (p=0.000 n=9+10)

name          old allocs/op  new allocs/op  delta
MemBuffer-10       120 ±10%       111 ±11%   -6.86%  (p=0.030 n=10+10)

```

The chunked implementation writes exactly the same # of events as the old implementation, which is about ~1179330 events.

Fixes: https://github.com/cockroachdb/cockroach/issues/84709
Fixes: https://github.com/cockroachdb/cockroach/issues/84582
(for now...)

jayshrivastava committed:
kvevent: refactor blocking buffer benchmark

This change makes several improvements to the blocking
buffer benchmark.

Firstly, it moves expensive setup operations such as span
generation to before b.ResetTimer which omits their stats
from the benchmark results.

Secondly, it moves stopProducers() such that the benchmark
will fill the buffer completely, then read out all the events.
Previously, producers could continue writing while events were
being read, which made benchmark results harder to interpret.

Finally, it adds accounting for the number of events/rows
written to the buffer. This helps verify that different
implementations of the buffer do not reduce its capacity.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85832)
<!-- Reviewable:end -->

jayshrivastava committed:
kevent: implement chunked buffer event queue

This change implements a simple chunked event queue.
The purpose of this queue is to be used by
kvevent.blockingBuffer in subsequent commits.

Release note: None

jayshrivastava committed:
kvevent: refactor memory buffer to chunked linked list

This change refactors kvevent/blocking_buffer.go to use
a chunked linked list instead of a regular linked list to
reduce pointer usage and memory allocations.

Please see the encapsulating PR
(https://github.com/cockroachdb/cockroach/pull/85832)
for benchmark results.

Release note: None

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85832)* status: :shipit: complete! 0 of 0 LGTMs obtained

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N97BclaAavBqk-TsN0d:-N97BclaAavBqk-TsN0e:b8qusw2) (no related file):*
Please let me know your thoughts about the slower performance. 

From some initial investigation, it looks like the the slowdown comes from copying events when we pop from the buffer. A possible fix would be to inline functions as much as possible at the cost of messier code. I'm also not sure if the slowdown is significant enough for it to matter in production.

```     
ROUTINE ======================== github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent.(*blockingBuffer).pop in github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go
Flat         Cum.  
…
1.39s        2.84s     84:   e, ok = b.mu.queue.dequeue()
...

ROUTINE ======================== github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent.(*bufferEventChunkQueue).dequeue in github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go
Flat         Cum.  
…
890ms      1.01s     45:   e, ok, consumedAllFromChunk := l.head.pop()
...

ROUTINE ======================== github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent.(*blockingBuffer).Get in github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go
Flat         Cum.  
...
780ms      3.24s    129:           got, ok, err := b.pop()
...

```
![Screen Shot 2022-08-10 at 10.51.06 AM.png](https://files.reviewableusercontent.io/images/reviewable/18633281/IjRpLIQuRKU8MJgiBfcC4WJ-/Screen%20Shot%202022-08-10%20at%2010.51.06%20AM.png)

___


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
> Reviewed 1 of 1 files at r1, 2 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85832)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jayshrivastava, and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N97BclaAavBqk-TsN0d:-N98ZZP52K95inQMUTQ3:b-24vsrq) (no related file):*
> From some initial investigation, it looks like the the slowdown comes from copying events when we pop from the buffer. A possible fix would be to inline functions as much as possible at the cost of messier code.

Writing this before I actually read the code so it's probably wrong but...could we make the function args more pointery (instead of inlining the functions) so there's less copying?

___
___
*[`pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N98_0lE5EzMUf6P4PAu:-N98_0lE5EzMUf6P4PAv:b-4547bc) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f02ce0d7cf185988b42256ca51bebedbf6e395/pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

Nit: 2022
___
*[`pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go` line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N98_r9W0ruT542813v9:-N98_r9W0ruT542813vA:b44af80) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f02ce0d7cf185988b42256ca51bebedbf6e395/pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go#L75)):*
> ```Go
> 
> type bufferEventChunk struct {
> 	events     [bufferEventChunkArrSize]Event
> ```

What happens if you make this a [bufferEventChunkArrSize]*Event instead, and changed all these methods to take and return a *Event? Does that increase allocations somehow or end up needing to be dereferenced anyway? Naively I'd expect it to be better.
___
*[`pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go` line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N98dpY-3Q3jSazZSgc8:-N98dpY-3Q3jSazZSgc9:b-w8ud7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f02ce0d7cf185988b42256ca51bebedbf6e395/pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go#L76)):*
> ```Go
> type bufferEventChunk struct {
> 	events     [bufferEventChunkArrSize]Event
> 	head, tail int32
> ```

So long as we're optimizing memory footprint might as well make these uint8.
___
*[`pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go` line 91 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N98cboZ2PhFT6CfQ4yB:-N98cboZ2PhFT6CfQ4yC:bpe7umd) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f02ce0d7cf185988b42256ca51bebedbf6e395/pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go#L91)):*
> ```Go
> 
> func freeBufferEventChunk(c *bufferEventChunk) {
> 	*c = bufferEventChunk{}
> ```

Haven't written a sync.Pool thing before but this looks wrong, won't this allocate a new bufferEventChunk? Would it be safe to instead set head and tail to 0? Which wouldn't be a memory leak because it's going back into the pool.
___
*[`pkg/ccl/changefeedccl/kvevent/chunked_event_queue_test.go` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N98eBE9Baij-72dgYbR:-N98eBEA6l_Vd6oD4ePo:bb6ht37) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f02ce0d7cf185988b42256ca51bebedbf6e395/pkg/ccl/changefeedccl/kvevent/chunked_event_queue_test.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

2022
___
*[`pkg/ccl/changefeedccl/kvevent/chunked_event_queue_test.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85832#-N98eH3332Jz8E27D_Q5:-N98eH3332Jz8E27D_Q6:brwlqvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f02ce0d7cf185988b42256ca51bebedbf6e395/pkg/ccl/changefeedccl/kvevent/chunked_event_queue_test.go#L31)):*
> ```Go
> 	assert.True(t, q.empty())
> 
> 	// Add events to fill 5 chunks and assume they are consumed in fifo order.
> ```

assert, not assume I think


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go:
```diff
@@ -0,0 +1,121 @@
+
+	e, ok, consumedAllFromChunk := l.head.pop()
+
+	if consumedAllFromChunk {
```
It would be nice if we didn't have to do these recursive calls.
I think the code (and datastructures) could be simplified.
Currently, we have:
```
type bufferEventChunkQueue struct {
	head, tail *bufferEventChunk
}
type bufferEventChunk struct {
	events     [bufferEventChunkArrSize]Event
	head, tail int32
	next       *bufferEventChunk // linked-list element
}
```

It's a bit sad that each chunk has to carry head/tail counters.
I think those counters can be lifted to bufferEventQueue
(drop the head/tail from bufferEventChunk):
```
type bufferEventChunkQueue struct {
  head, tail *bufferEventChunk
  headChunkIdx, tailChunkIdx int      
}
```

Enqueue becomes pretty easy:
```
func (l *bufferEventChunkQueue) enqueue(e Event) {
   if l.tail == nil || l.tailChunkIdx == bufferEventChunkArrSize {
      // Add new chunk; this either creates or updates l.tail
   }
   l.tail.events[l.tailChunkIdx] = e
   l.tailChunkIdx++
}
```

Dequeue is similarly easy
```
func (l *bufferEventChunkQueue) dequeue() (e Event, ok bool) {
   // You always dequeue from head
   e = l.head.events[l.headChunkIdx]
  l.headChunkIdx++
  if l.headChunkIdx == bufferEventChunkArrSize {
     // free l.head
  }
}
```

jayshrivastava commented:
Closing because this PR is outdated and will be messy to update. Will link the new PR once it's up.
